packages/stream-receiver/src/index.ts

```diff
@@ -0,0 +1,430 @@
+
+      if (hasPaymentTag) {
+        details.paymentTag = reader.readVarOctetString().toString('ascii')
+      }
+
+      if (hasReceiptDetails) {
+        details.receiptSetup = {
+          nonce: reader.read(16),
+          secret: reader.read(32),
+        }
+      }
+
+      if (hasAssetDetails) {
+        details.asset = {
+          code: reader.readVarOctetString().toString(),
+          scale: reader.readUInt8Number(),
+        }
+      }
+
+      return details
+    } catch (_) {
+      // No-op: failed decryption or structurally invalid
+    }
+  }
+
+  /**
+   * Extract the `paymentTag` from the given destination ILP address, or return `undefined`
+   * if the connection token is invalid or no payment tag was encoded.
+   */
+  decodePaymentTag(destinationAddress: string): string | void {
+    const token = this.extractTokenFromAddress(destinationAddress)
+    if (token) {
+      return this.decryptToken(token)?.paymentTag
+    }
+  }
+
+  /**
+   * Validate and decrypt the destination ILP address of an incoming ILP Prepare:
+   * ensure it's addressed to the server and decode encrypted metadata to attribute and handle the payment.
+   */
+  createReply(prepare: IlpPrepare): IncomingMoney | IlpReply {
+    const connectionId = sha256(Buffer.from(prepare.destination, 'ascii')).toString('hex')
+    const log = createLogger(`ilp-receiver:${connectionId.slice(0, 6)}`)
+    const reply = new ReplyBuilder().setIlpAddress(this.serverAddress)
+
+    // Ensure the packet is addressed to us
+    const token = this.extractTokenFromAddress(prepare.destination)
+    if (!token) {
+      log.trace('got packet not addressed to the receiver. destination=%s', prepare.destination)
+      return reply.buildReject(IlpError.F02_UNREACHABLE)
+    }
+
+    const connectionDetails = this.decryptToken(token)
+    if (!connectionDetails) {
+      log.trace('invalid connection token: cannot attribute incoming packet. token=%s', token)
+      return reply.buildReject(IlpError.F06_UNEXPECTED_PAYMENT)
+    }
+    const { paymentTag, receiptSetup, asset } = connectionDetails
+
+    log.debug('got incoming Prepare. amount: %s', prepare.amount)
+
+    const sharedSecret = hmac(this.sharedSecretKeyGen, token)
+    const encryptionKey = hmac(sharedSecret, StreamServer.ENCRYPTION_KEY_STRING)
+    let streamRequest: Packet
+    try {
+      streamRequest = Packet._deserializeUnencrypted(decrypt(encryptionKey, prepare.data))
+    } catch (_) {
+      log.trace('rejecting with F06: failed to decrypt STREAM data') // Inauthentic, could be anyone
+      return reply.buildReject(IlpError.F06_UNEXPECTED_PAYMENT)
+    }
+
+    if (+streamRequest.ilpPacketType !== IlpPacketType.Prepare) {
+      log.warn('rejecting with F99: invalid STREAM packet type') // Sender violated protocol, or intermediaries swapped valid STREAM packets
+      return reply.buildReject(IlpError.F00_BAD_REQUEST) // Client should not retry, but don't include STREAM data since the request was inauthentic
```

Comment:
Nice catch, intended this to be F00 since it if gets back to the client, they definitely should not retry (but it shouldn't include a STREAM packet since it wasn't authentic). I'll fix the log line.

Reply:
fyi [ilp-protocol-stream](https://github.com/interledgerjs/ilp-protocol-stream/blob/7ad483f5fd1a1d1e4dc58d7eef6a437594646260/src/connection.ts#L538-L541) returns a `F06: Unexpected Payment` for this case. I don't think it's specified, though.

Reply:
https://github.com/interledgerjs/interledgerjs/compare/b6a4203...ko/stream-receiver#diff-83d26dbe15dbfbf4cec69b03634e44acR323